### PR TITLE
Support polygon mask training

### PR DESCRIPTION
**Pull request by yhcao6**
Add polygon mask training support to speed up mask preprocessing and mask target computation.
Refactor mask transforms, add two structures BitMapMasks and PolygonMasks.

### CODE REVIEW

File: mmdet/datasets/structures.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
+import mmcv
+import numpy as np
+import pycocotools.mask as maskUtils
+import torch
+
+
```
Comment (ZwwWayne) [inline mmdet/datasets/structures.py]: It may be better to move data structures into mmdet.cores since these structures are commonly used not only in data pipeline.

File: mmdet/datasets/structures.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+class BitMapMasks:
```
Comment (hellock) [inline mmdet/datasets/structures.py]: Add a docstring.

File: mmdet/datasets/structures.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
+import mmcv
+import numpy as np
+import pycocotools.mask as maskUtils
+import torch
```
Comment (hellock) [inline mmdet/datasets/structures.py]: Add a base class for instance masks to define the interfaces.

File: mmdet/core/mask/structures.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+    Args:
+        masks (ndarray): ndarray of masks in shape (N, H, W), where N is
+            the number of objects.
+        height (int): height of masks
+        width (int): width of masks
+    """
+
+    def __init__(self, masks, height, width):
+        self.height = height
+        self.width = width
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: Are `height` and `width` necessary? They can be obtained from `masks` even if the shape of `masks` is (0, h, w).

File: mmdet/core/mask/structures.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+    def __len__(self):
+        return len(self.masks)
+
+    def rescale(self, scale, interpolation='nearest'):
+        """Rescale masks as large as possible while keeping the aspect ratio.
+        For details can refer to `mmcv.imrescale`
+
+        Args:
+            scale (tuple[int]): the maximum size
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: Specify if it is (h, w) or (w, h).

File: mmdet/core/mask/structures.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+class Masks(object):
+
+    def rescale(self, scale, interpolation='nearest'):
+        pass
+
+    def resize(self, h, w, interpolation='nearest'):
+        pass
+
+    def flip(self, flip_direction):
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: Set a default value.

File: mmdet/core/mask/structures.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+
+    def crop(self, bbox):
+        """Crop each mask by the given bbox.
+
+        Args:
+            bbox (ndarray): bbox in format [x1, y1, x2, y2], shape (4, )
+
+        Return:
+            BitMapMasks: the cropped masks.
+        """
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: Check the input arguments.

File: mmdet/core/mask/structures.py [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+        pass
+
+    def crop(self, bbox):
+        pass
+
+    def crop_and_resize(self,
+                        bboxes,
+                        mask_h,
+                        mask_w,
+                        assigned_inds,
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: Rename it to `inds`. The interface of masks should not be correlated to the assignment.

File: mmdet/core/mask/structures.py [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+    def pad(self, h, w, pad_val):
+        pass
+
+    def crop(self, bbox):
+        pass
+
+    def crop_and_resize(self,
+                        bboxes,
+                        mask_h,
+                        mask_w,
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: `crop_and_resize(bboxes, out_shape, ...)`

File: mmdet/core/mask/structures.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+    def rescale(self, scale, interpolation='nearest'):
+        pass
+
+    def resize(self, h, w, interpolation='nearest'):
+        pass
+
+    def flip(self, flip_direction):
+        pass
+
+    def pad(self, h, w, pad_val):
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: `pad(self, out_shape, pad_val)`

File: mmdet/core/mask/structures.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+import pycocotools.mask as maskUtils
+import torch
+
+
+class Masks(object):
+
+    def rescale(self, scale, interpolation='nearest'):
+        pass
+
+    def resize(self, h, w, interpolation='nearest'):
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: `resize(self, out_shape, interpolation='nearest')`

File: mmdet/core/mask/structures.py [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+
+        Args:
+            bboxes (ndarray): bboxes in format [x1, y1, x2, y2], shape (N, 4)
+            mask_h (int): height of resized mask
+            mask_w (int): width of resized mask
+            assigned_inds (ndarray): indexes to assign masks to each bbox
+            interpolation (str): see `mmcv.imresize`
+
+        Return:
+            ndarray: the cropped and resized masks.
```
Comment (hellock) [inline mmdet/core/mask/structures.py]: To keep the consistency, return a BitmapMask.

File: tests/test_masks.py [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+    flipped_masks = polygon_masks.flip(flip_direction='horizontal')
+    assert len(flipped_masks) == 0
+
+    # horizontally flip with polygon masks contain 3 instances
+    raw_masks = dummy_raw_polygon_masks((3, 28, 28))
+    polygon_masks = PolygonMasks(raw_masks, 28, 28)
+    flipped_masks = polygon_masks.flip(flip_direction='horizontal')
+    flipped_flipped_masks = flipped_masks.flip(flip_direction='horizontal')
+    assert (polygon_masks.to_ndarray() == flipped_flipped_masks.to_ndarray()
+            ).all()
```
Comment (hellock) [inline tests/test_masks.py:324]: This is not safe. A better way is to check the flipped masks directly. `assert polygon_masks.to_ndarray() == flipped_masks.to_ndarray()[:, :, ::-1]`

File: mmdet/datasets/pipelines/loading.py [hunk 14]
```diff
@@ -128,11 +129,30 @@ def _poly2mask(self, mask_ann, img_h, img_w):
         mask = maskUtils.decode(rle)
         return mask
 
+    def process_polygons(self, polygons):
+        """ Convert polygons to ndarray and filter invalid polygons.
+        Args:
+            polygons (list of list): polygons of one instance.
+        Returns:
+            polygons (list of ndarray).
+        """
```
Comment (hellock) [inline mmdet/datasets/pipelines/loading.py:138]: Add a blank line between two blocks of the docstring.

File: mmdet/datasets/pipelines/transforms.py [hunk 15]
```diff
@@ -868,6 +814,11 @@ def __call__(self, results):
             if self.filter_lost_elements:
                 results['idx_mapper'] = np.arange(len(results['bboxes']))
 
+        # TODO: Support mask structure in albu
+        if 'masks' in results:
+            ori_masks = results['masks']
```
Comment (hellock) [inline mmdet/datasets/pipelines/transforms.py:819]: What if ori_masks is PolygonMasks?
